### PR TITLE
MSC2176: Update the redaction rules

### PR DESCRIPTION
**Pull request by richvdh**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/rav/proposals/update-redaction/proposals/2176-update-redaction-rules.md).

### CODE REVIEW

File: proposals/2176-update-redaction-rules.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+The following *event* keys should be *removed* from the list of those to be
+preserved by a redaction:
+
+ * `membership`
+ * `prev_state`
+
+(Note this refers to the *event-level* `membership` property, rather than the
+similarly-named sub-property under the `content` key.)
+
+Ratinale: neither of the above properties have defined meanings in the Matrix
```
Comment (ara4n) [inline proposals/2176-update-redaction-rules.md]: Rationalel

**Comment by ara4n** [comment]
looks plausible to me. sounds like we should get it out there in a room v6 sooner than later

**Comment by richvdh** [comment]
We probably need to get agreement on #1996 and #1601 as a prerequisite here. If people have opinions on them, perhaps they could weigh in on the relevant bugs.

**Comment by bwindels** [comment]
For completeness, there is also some [ongoing discussion as part of MSC1849/2154](https://github.com/matrix-org/matrix-doc/pull/2154#discussion_r300725473) about preserving `content/m.relates_to/rel_type` on redaction for at least edits, if not all relations.

File: proposals/2176-update-redaction-rules.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+hashes](https://matrix.org/docs/spec/server_server/r0.1.2#calculating-the-reference-hash-for-an-event)
+(and hence event IDs) are calculated.
+
+The following *event* keys should be *removed* from the list of those to be
+preserved by a redaction:
+
+ * `membership`
+ * `prev_state`
+
+(Note this refers to the *event-level* `membership` property, rather than the
```
Comment (babolivier) [inline proposals/2176-update-redaction-rules.md]: Do we actually have an event-level `membership` property in events? I found nothing of the sort in https://matrix.org/docs/spec/client_server/r0.5.0#event-fields (nor in the S2S spec)

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: well no, that's the point, and why I believe it should be removed from the list of things to be preserved by a redaction :)

Comment (babolivier) [reply proposals/2176-update-redaction-rules.md]: Ah, gotcha, I didn't understand right away what it was supposed to be removed from (nothing to do with the proposal, just my brain malfunctioning)

File: proposals/2176-update-redaction-rules.md (hunk 1 above)
Comment (KitsuneRal) [inline proposals/2176-update-redaction-rules.md]: Rationale, even :)
```suggestion
Rationale: neither of the above properties have defined meanings in the Matrix
```

File: proposals/2176-update-redaction-rules.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+ * `m.room.power_levels` should allow the `notifications` key. Rationale:
+   symmetry with the other `power_levels` settings. (Maybe? See
+   https://github.com/matrix-org/matrix-doc/issues/1601#issuecomment-511237744.)
+
+
+## Potential issues
+
+What if there is spam in sub-properties of the `notifications` property of
+power-levels? Should we not be able to redact it?
```
Comment (KitsuneRal) [inline proposals/2176-update-redaction-rules.md]: Sounds exactly like a way to attach a spam dumbbell to a data structure that happened with GPG folks recently. Though much less killing it's no less undesirable.

File: proposals/2176-update-redaction-rules.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+ * `m.room.redaction` should allow the `redacts` key (assuming
+   [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) is merged).
+   Rationale: currently, redacting a redaction can lead to inconsistent results
+   among homservers, depending on whether they receive the `m.room.redaction`
+   result before or after it is redacted (and therefore may or may not redact
+   the original event).
+
+ * `m.room.create` should allow the `room_version` key. Currently, redacting an
+   `m.room.create` event will make the room revert to a v1 room.
```
Comment (uhoreg) [inline proposals/2176-update-redaction-rules.md]: Also `m.federate` and `predecessor`?

File: proposals/2176-update-redaction-rules.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+ * `m.room.power_levels` should allow the `notifications` key. Rationale:
```
Comment (uhoreg) [inline proposals/2176-update-redaction-rules.md]: also `invite`?

File: proposals/2176-update-redaction-rules.md (hunk 3 above)
Comment (uhoreg) [reply proposals/2176-update-redaction-rules.md]: `notifications` doesn't seem any more abusable to me than the other keys in `m.room.power_levels` that we also keep.

Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: I'd argue it be kept so that users aren't likely to forget about it when they redact the event (although why would you redact the power levels in the first place?) - if it's the only key excluded, it can be easily missed. However, there is argument that it doesn't help the protocol at all to keep or throw it away, and so we shouldn't bother saving it.

The rationale behind why someone might redact a power level event would be interesting, I think. It could reveal what sorts of keys we should be keeping (from a convenience perspective) and whether it needs to be kept at all.

File: proposals/2176-update-redaction-rules.md (hunk 5 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: +1. Never realised that was missing.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: fixed by d324cac.

File: proposals/2176-update-redaction-rules.md (hunk 3 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: A solution to spam in the particular case of `notifications` is to specify that only the `room` sub-key be kept. Obviously that's not a general solution though.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: > users aren't likely to forget about it when they redact the event 

? You mean client implementors? It would be odd for a user to redact an event manually.

**Comment by richvdh** [comment]
A quick note on some of my thinking here.

As I see it, there is a scale between properties which we absolutely *must* preserve on redaction, for fear of breaking the protocol (mostly those involved in event auth), and those which have very little effect on the protocol at all so can safely be redacted. Somewhere in the middle we have things like `predecessor`, which probably isn't really a thing you want to redact, but actually the protocol will mostly keep working.

I'm trying quite hard to limit the list of things that get preserved to things which completely break the protocol, rather than just things that are a bit annoying/surprising when you redact them, mostly to keep the server-server protocol as simple as possible. The fact that we have to bump the room version to change the redaction rules also means that it's really not great to have to keep adding new rules. Rather, I see it as the job of either clients (or, possibly, the C-S API) to protect you from foot-guns.

All that said, there is room for debate on where different keys sit on that scale, and where we draw the line, but I'd like each candidate to be retained to be argued on its own merits.

File: proposals/2176-update-redaction-rules.md (hunk 4 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: I'm starting to wonder if it would make more sense to just say you can't redact an `m.room.create` event (or rather, that attempting to do so has no effect, so all keys are included in the reference hash).

What does it even mean to redact a create event?

File: proposals/2176-update-redaction-rules.md (hunk 3 above)
Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: I mean more when a user redacts the power level event: if they are redacting to clean up some typos then they might forget to restore their custom `notifications` settings. If they're redacting to reset things to default, then meh, whatever. If they're doing it for some other reason, do we care to try and help by keeping it around?

File: proposals/2176-update-redaction-rules.md (hunk 4 above)
Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: I'd heavily vote for not allowing it to be redacted. The creation event only has behaviour if it has `content` (otherwise it's role is simply being the first event in the room), and redacting a state event is not supposed to change the behaviour of the room (at least according to https://github.com/matrix-org/matrix-doc/pull/2032)

File: proposals/2176-update-redaction-rules.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+   * the `invite` key. Rationale: this is required to authenticate
+     `m.room.member` events with the `invite` membership. Currently, redacting
+     a `power_levels` event will mean that such events cannot be authenticated,
+     potentially leading to a split-brain room.
+
+   * the `notifications` key. Rationale: symmetry with the other `power_levels`
+     settings. (Maybe? See
+     https://github.com/matrix-org/matrix-doc/issues/1601#issuecomment-511237744.)
+
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]: `m.room.encryption`'s `algorithm` should be preserved.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: I disagree. It doesn't change the behaviour of the room at the server-server level.

File: proposals/2176-update-redaction-rules.md (hunk 3 above)
Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: Typos in power levels?.. Seems far-fetched a bit.

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: But if we eventually have two algorithms rather than one how should a client treat an event with redacted algorithm?

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: > But if we eventually have two algorithms rather than one

I don't think there is a way to end up with two encryption algorithms in the room state?

> how should a client treat `m.room.encryption` with redacted algorithm?

The same way that they should treat a new `m.room.encryption` event which disables encryption: with great suspicion.

Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: How are new clients to the room meant to do encryption if they don't know the algorithm? 


Can we please relax the restrictions on what goes into the algorithm? It affects client authors too, and it's ridiculous to exclude something because it doesn't affect the core protocol operation.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: > How are new clients to the room meant to do encryption if they don't know the algorithm?

They are not. But I don't understand what adding `algorithm` to the redaction-preservation list buys us, since we have the same problem if you send a new `m.room.encryption` event which disables encryption, which is IMHO a much bigger problem than redactions.

> Can we please relax the restrictions on what goes into the algorithm? It affects client authors too, and it's ridiculous to exclude something because it doesn't affect the core protocol operation.

I assume you mean the *redaction* algorithm here? Sorry, I believe it's important to keep the list as short as possible.

Comment (uhoreg) [reply proposals/2176-update-redaction-rules.md]: I'm inclined to agree with Travis in wanting to preserve the `algorithm` in `m.room.encryption` events, but given that we still have a problem of sending a new `encryption` event, I think it may be better to leave it out for now and tackle both issues in a separate MSC.

File: proposals/2176-update-redaction-rules.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+   `m.room.create` event will make the room revert to a v1 room.
+
+ * `m.room.power_levels` should allow:
+
+   * the `invite` key. Rationale: this is required to authenticate
+     `m.room.member` events with the `invite` membership. Currently, redacting
+     a `power_levels` event will mean that such events cannot be authenticated,
+     potentially leading to a split-brain room.
+
+   * the `notifications` key. Rationale: symmetry with the other `power_levels`
```
Comment (uhoreg) [inline proposals/2176-update-redaction-rules.md]: Should we also be adding `notifications` too the auth rules so that they're checked in the same way to how the other keys are checked (10.c under https://matrix.org/docs/spec/rooms/v1#authorization-rules)?

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: probably, but ugh.

**Comment by richvdh** [comment]
I've updated the MSC to (1) change the create event so that all keys are preserved, and (2) clarify that I don't think the `notifications` property should be preserved, and (2b) to record why I think this is important.

Further input welcome...

File: proposals/2176-update-redaction-rules.md (hunk 4 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: done, as of 9e264fe

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (erikjohnston) [reply proposals/2176-update-redaction-rules.md]: One of the big motivations for redactions is to allow stripping as much user defined data as possible from an event, without breaking the room in such a way other servers cannot join/interact with it. This was primarily to protect against users adding illegal content to events and servers having no choice but to throw away the room to get rid of it from the DB, though the scope of redactions has increased significantly since.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: @erikjohnston: yeeesss... I'm not quite clear what your point is though?

File: proposals/2176-update-redaction-rules.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+The following *event* keys should be *removed* from the list of those to be
+preserved by a redaction:
+
+ * `membership`
+ * `prev_state`
+
+(Note this refers to the *event-level* `membership` property, rather than the
+similarly-named sub-property under the `content` key.)
+
+Rationale: neither of the above properties have defined meanings in the Matrix
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]: Historically they did, if someone were to go digging far enough back:
```suggestion
Rationale: neither of the above properties have defined meanings anymore in the Matrix
```

File: proposals/2176-update-redaction-rules.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+protocol, so there is no reason for them to be special-cased in this way.
+
+The following should be added to the list of subkeys of the content property
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
The following are to be added to the list of subkeys of the content property
```

File: proposals/2176-update-redaction-rules.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+which should be preserved:
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
which are preserved:
```

File: proposals/2176-update-redaction-rules.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+ * `m.room.create` should preserve *all* content. Rationale: the values in a
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
 * `m.room.create` preserves *all* content. Rationale: the values in a
```


... and repeat for the rest of the proposal.

File: proposals/2176-update-redaction-rules.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+   event.
+
+ * `m.room.redaction` should allow the `redacts` key (assuming
+   [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) is merged).
+   Rationale: currently, redacting a redaction can lead to inconsistent results
+   among homservers, depending on whether they receive the `m.room.redaction`
+   result before or after it is redacted (and therefore may or may not redact
+   the original event).
+
+ * `m.room.power_levels` should allow:
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]: can we stick some words in here to remind the reader that this list is *adding* to the existing rules and not replacing? After the first two points with some minor context switches, it's easy to forget that this list is additive.

File: proposals/2176-update-redaction-rules.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+     potentially leading to a split-brain room.
+
+## Other properties considered for preservation
+
+Currently it is *not* proposed to add these to the list of properties which are
+proposed for a redaction:
+
+ * The `notifications` key of `m.room.power_levels`. Unlike the other
+   properties in `power_levels`, `notifications` does not play a part in
+   authorising the events in the room graph. Once the `power_levels` are
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]: It's true the event graph doesn't care, however the historical context of rooms in clients will be affected by this decision. Currently if one pings the room and redacts the power levels, the client has no choice but to de-highlight the notification. This is less relevant for push notifications as they are already executed prior to the power level event being redacted (it'd be weird and unlikely that someone pings the room then immediately redacts the power levels), however when the user is scrolling back through the room or searching they will have lost highlighting on the event.

tldr: we should preserve it for the purposes of historically accurate highlighting of the room.

File: proposals/2176-update-redaction-rules.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
+# MSC2176: Update the redaction rules
```
Comment (turt2live) [inline proposals/2176-update-redaction-rules.md]: Should `m.room.guest_access` preserve `guest_access`? If it doesn't, what does it mean to redact a guest access policy of `can_join`? Ignoring the fact that redacting most state events violates the schemas (`name` for instance is required on the room name event), the guest access event defaults to `forbidden`: are all guests kicked from the room now? How does a remote server even validate that a guest account is a guest account?

I believe this is the only state event which raises questions about authorization: the remainder are already covered by the proposal or the existing algorithm.

File: proposals/2176-update-redaction-rules.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+similarly-named sub-property under the `content` key.)
+
+Rationale: neither of the above properties have defined meanings in the Matrix
+protocol, so there is no reason for them to be special-cased in this way.
+
+The following should be added to the list of subkeys of the content property
+which should be preserved:
+
+ * `m.room.create` should preserve *all* content. Rationale: the values in a
+   `create` event are deliberately intented to last the lifetime of the room,
```
Comment (richvdh) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
   `create` event are deliberately intended to last the lifetime of the room,
```

File: proposals/2176-update-redaction-rules.md (hunk 13 above)
Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: Is there any difference to redacting, say, a room topic? It also affects user interface, and clients should explicitly process that, and few actually do.

File: proposals/2176-update-redaction-rules.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+   authorising the events in the room graph. Once the `power_levels` are
+   replaced, historical values of the `notifications` property are
+   irrelevant. There is therefore no need for it to be protected from
+   redactions.
+
+ * The `algorithm` key of `m.room.encryption`. Again, historical values of
+   `m.room.encryption` have no effect, and servers do not use the value of the
+   property to authenticate events.
+
+   The effect of redacting an `m.room.redaction` event is much the same as that
```
Comment (KitsuneRal) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
   The effect of redacting an `m.room.encryption` event is much the same as that
```

File: proposals/2176-update-redaction-rules.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+   of sending a new `m.room.redaction` event with no `algorithm` key. It's
```
Comment (KitsuneRal) [inline proposals/2176-update-redaction-rules.md]:
```suggestion
   of sending a new `m.room.encryption` event with no `algorithm` key. It's
```

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: I still don't quite understand what a client should do if it loads a room (having no previous state) and receives m.room.encryption with empty algorithm (right away, again, no previous state). Should it consider the room encrypted at all? If yes, which algorithm should it assume?

File: proposals/2176-update-redaction-rules.md (hunk 13 above)
Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: The slight difference is that the topic should absolutely be redactable for when someone abuses the text field as a means for sending abuse. The redaction algorithm is also meant to give people the ability to remove harmful content/messages, however whether a message is highlighted or not doesn't really change that.

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: @KitsuneRal this isn't a question that is limited to redactions, but: I would suggest that a client warns a user that the room state is invalid.

File: proposals/2176-update-redaction-rules.md (hunk 7 above)
Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: ftr this is now a dedicated issue: https://github.com/matrix-org/matrix-doc/issues/2198

For simplicity, this can be its own MSC I think.

File: proposals/2176-update-redaction-rules.md (hunk 14 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: remote servers can't distinguish between a guest account and a regular user account, so `m.room.guest_access` isn't involved in the auth checks done on events received over federation (see also: https://matrix.org/docs/spec/rooms/v1#authorization-rules). I'd therefore argue that it's less important to preserve its value than it is for `power_levels` and friends.

To answer your other questions:

> If it doesn't, what does it mean to redact a guest access policy of `can_join`....  the guest access event defaults to forbidden: are all guests kicked from the room now?

according to https://matrix.org/docs/spec/client_server/r0.5.0#id159: yes, although that certainly isn't implemented in synapse currently. (It's also not implemented when the guest access changes due to state resolution rather than due to a new event).

> Ignoring the fact that redacting most state events violates the schemas

yeah; this is a general problem. It begins to interact with validation of events in general (see also: https://github.com/matrix-org/matrix-doc/issues/1646): the current situation is that there is no verification that state events received over federation match the schema, so we have to be prepared to handle events that do not match the schema anyway.

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: I'm going to resolve this thread; if people feel strongly that I'm making the wrong call on this, then reopen/start a new one.

File: proposals/2176-update-redaction-rules.md (hunk 13 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: I'm not sure this is a strong enough argument for me to want to include it in the list. I'll accept lobbying though...

File: proposals/2176-update-redaction-rules.md (hunk 12 above)
Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: clarification added

**Comment by richvdh** [comment]
Nothing else seems to be happening here, so I'd like to propose FCP:

@mscbot fcp merge

File: proposals/2176-update-redaction-rules.md (hunk 13 above)
Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: My lobbying budget seems to have run out for the quarter :(

If the arguments above aren't convincing enough, there's nothing new I can provide.

File: proposals/2176-update-redaction-rules.md (hunk 6 above)
Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: I thought that I sent this comment but actually it got stuck behind the unpressed "Submit review" button (and then GitHub tucked it back where I drafted it. Hope it's placed properly now.

I think I understand where the disconnect is: I'm afraid I don't quite buy the argument that for the sake of reducing the change surface we're restricting redaction exclusions to server interaction pieces. It's the clients ultimately sending and using those events, so redaction should at least somewhat protect basic room integrity from the client perspective as well.

Now the state can be broken in a lot of different ways and (I totally agree here) clients ought to warn users about things that may break their experience; but when it comes to m.room.encryption, there's no good way for a client to handle any readable content of the room once that state is broken.

To reiterate, the case in point is not about repeated m.room.encryption events; it's about a client joining/loading a room where m.room.encryption is already redacted. From the client perspective, this room is irreparably broken along with all of its encrypted messages already in DAG.

Going as far as imagining a m.room.encryption event with obscene/illegal contents for "algorithm" sent to the previously unencrypted room (again - repeated encryption events are not proper and I don't consider them here), in that case the room also gets broken for everyone but no user content is lost - because the room was unencrypted and now it is "encrypted" with non-sensical algorithm. In such case it would be great to have algorithm redacted even though the room has to be thrown out, in order to eliminate improper content from DAG. Ideally such events should not be accepted to DAG in the first place - then the case for the breakage and redaction doesn't emerge.

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: @KitsuneRal: I think there might be a misunderstanding here.

> there's no good way for a client to handle any readable content of the room once that state is broken
> ...
> this room is irreparably broken along with all of its encrypted messages already in DAG.

`m.room.encryption` is not supposed to define how you interpret received events, so I don't agree that the existing messages are invalidated once `m.room.encryption` is redacted. 

>  again - repeated encryption events are not proper

there is nothing wrong with repeated `m.room.encryption` events?

Comment (KitsuneRal) [reply proposals/2176-update-redaction-rules.md]: > `m.room.encryption` is not supposed to define how you interpret received events, so I don't agree that the existing messages are invalidated once `m.room.encryption` is redacted.

That's a pretty strong statement. Then what's the purpose of `algorithm` there?

> there is nothing wrong with repeated `m.room.encryption` events?

Except they should be treated with great suspicion because there's no defined semantics for any `m.room.encryption` event except the first one?

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: > Then what's the purpose of algorithm there?

It indicates to clients how they should encrypt messages in this room.

> there's no defined semantics for any m.room.encryption event except the first one?

I don't think that's true. I can't find anything in the spec about this, but my mental model is that clients should keep a record of rooms which are encrypted (along with the encryption algorithm), and give the user a warning if they send a message in a room which was previously encrypted and is no longer (or if the encryption is 'weaker', for some handwavey definition of weaker).

But certainly it's valid for there to be multiple encryption events in a room, changing minor settings.

Comment (turt2live) [reply proposals/2176-update-redaction-rules.md]: There's also the problem of clients joining a room which has a redacted encryption event: They won't have the prior knowledge that it was encrypted and will assume it's not.

ftr riot-web does exactly as you describe: it ignores changes to the encryption event it wasn't expecting.

Comment (uhoreg) [reply proposals/2176-update-redaction-rules.md]: I've created https://github.com/matrix-org/matrix-doc/issues/2272 for clarifying the semantics of `m.room.encryption` events.  I'd suggest ignoring `m.room.encryption` for the purposes of this MSC, and discussing in that issue for a future MSC, as I think that it's a larger issue that shouldn't hold this one up.

**Comment by richvdh** [comment]
It seems like this fixes most of https://github.com/matrix-org/synapse/issues/1831, but doesn't handle `third_party_invite` in `m.room.membership`. I vaguely remember considering that and dismissing it, but do not recall any futher details :/

File: proposals/2176-update-redaction-rules.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
+# MSC2176: Update the redaction rules
```
Comment (uhoreg) [inline proposals/2176-update-redaction-rules.md]: i just noticed that the redaction algorithm doesn't care about whether events are state events are not.  Is this something that we need to care about?

Comment (richvdh) [reply proposals/2176-update-redaction-rules.md]: possibly. Given the amount of "discussion" which arose from the MSC so far, I've come to the conclusion that having a single MSC to fix all known problems with redaction is misguided - in other words, let's punt that?

Comment (uhoreg) [reply proposals/2176-update-redaction-rules.md]: Fair enough.  I've opened https://github.com/matrix-org/matrix-doc/issues/2308 to track this.

**Comment by anoadragon453** [comment]
@richvdh Does this proposal have an implementation as of yet?

**Comment by richvdh** [comment]
not yet

**Comment by clokep** [comment]
> Does this proposal have an implementation as of yet?

There's now an implementation at matrix-org/synapse#8984.

File: proposals/2176-update-redaction-rules.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+ * The `notifications` key of `m.room.power_levels`. Unlike the other
+   properties in `power_levels`, `notifications` does not play a part in
+   authorising the events in the room graph. Once the `power_levels` are
+   replaced, historical values of the `notifications` property are
+   irrelevant. There is therefore no need for it to be protected from
+   redactions.
+
+ * The `algorithm` key of `m.room.encryption`. Again, historical values of
+   `m.room.encryption` have no effect, and servers do not use the value of the
+   property to authenticate events.
```
Comment (erkinalp) [inline proposals/2176-update-redaction-rules.md:65]: This might be useful for backfilling into E2EE rooms.

**Comment by richvdh** [comment]
Current state of this MSC is that there is an implementation (over at matrix-org/synapse#8984), but it somewhat relies on an implementation of [MSC2174](https://github.com/matrix-org/matrix-doc/pull/2174) (see [thread](https://github.com/matrix-org/synapse/pull/8984#discussion_r549397543)) so we don't want to incorporate it into a room version until we have also implemented MSC2174.

**Comment by turt2live** [comment]
This MSC is currently assigned to v10, though apparently we need to unblock MSC2174 too. https://github.com/matrix-org/matrix-spec-proposals/pull/3604

**Comment by turt2live** [comment]
This went back to not having an assigned room version (pulled out of v10).

**Comment by turt2live** [comment]
This is now formally assigned to room version 11: https://github.com/matrix-org/matrix-spec-proposals/pull/3820

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1604

**Comment by turt2live** [comment]
Merged 🎉
